juju/machine.py

```diff
@@ -235,10 +235,20 @@ def dns_name(self):
         May return None if no suitable address is found.
         """
         addresses = self.safe_data['addresses'] or []
-        for address in addresses:
+        return_addresses = []
+        ordered_addresses = []
+        ordered_scopes = ['public', 'local-cloud', 'local-fan']
+        for scope in ordered_scopes:
+            for address in addresses:
+                if scope == address['scope']:
+                    ordered_addresses.append(address)
```

Comment:
Since you're now looking for the scopes in order, and exhausting the addresses for each, you could return the first one you find here and drop the rest up until `return None`.

Happy to approve then.